scheduler/replica_scheduler.go

```diff
@@ -286,6 +292,173 @@ func (rcs *ReplicaScheduler) scheduleReplicaToDisk(replica *longhorn.Replica, di
 		replica.Name, replica.Spec.NodeID, replica.Spec.DiskID, replica.Spec.DataPath)
 }
 
+// CheckAndReuseFailedReplica will check:
+//     1. If there is currently reusable failed replica
+//     2. If there is potential reusable failed replica
+func (rcs *ReplicaScheduler) CheckAndReuseFailedReplica(replicas map[string]*longhorn.Replica, volume *longhorn.Volume, hardNodeAffinity string) (bool, *longhorn.Replica, error) {
+	nodesInfo, err := rcs.getNodeInfo()
+	if err != nil {
+		return false, nil, err
+	}
+
+	hasPotentialReuseableReplica := false
+	failedReplicas := map[string]*longhorn.Replica{}
+	for _, r := range replicas {
+		isReusable, isPotentialReuseableReplica := rcs.checkFailedReplicaReusability(r, volume, nodesInfo, hardNodeAffinity)
+		if isPotentialReuseableReplica {
+			hasPotentialReuseableReplica = true
+		}
+		if !isReusable {
+			continue
+		}
+		if existingReplica, exists := failedReplicas[r.Spec.NodeID]; exists {
+			latestReplica := GetLatestFailedReplica(r, existingReplica)
+			failedReplicas[latestReplica.Spec.NodeID] = latestReplica
+		} else {
+			failedReplicas[r.Spec.NodeID] = r
+		}
+	}
+
+	res, err := rcs.chooseReusableReplica(nodesInfo, replicas, failedReplicas, volume)
+	return hasPotentialReuseableReplica, res, err
+}
+
+func (rcs *ReplicaScheduler) checkFailedReplicaReusability(r *longhorn.Replica, v *longhorn.Volume, nodeInfo map[string]*longhorn.Node, hardNodeAffinity string) (bool, bool) {
+	if r.Spec.FailedAt == "" {
+		return false, false
+	}
+	if r.Spec.NodeID == "" || r.Spec.DiskID == "" {
+		return false, false
+	}
+	if r.Spec.RebuildRetryCount >= FailedReplicaMaxRetryCount {
+		return false, false
+	}
+	if r.Status.EvictionRequested {
+		return false, true
+	}
+	if hardNodeAffinity != "" && r.Spec.NodeID != hardNodeAffinity {
+		return false, true
+	}
+
+	node, exists := nodeInfo[r.Spec.NodeID]
+	if !exists {
+		return false, true
+	}
+	diskSpec, exists := node.Spec.Disks[r.Spec.DiskID]
+	if !exists {
+		return false, true
+	}
+	diskStatus, exists := node.Status.DiskStatus[r.Spec.DiskID]
+	if !exists {
+		return false, true
+	}
+
+	diskReadyCondition := types.GetCondition(diskStatus.Conditions, types.DiskConditionTypeReady)
+	diskSchedulableCondition := types.GetCondition(diskStatus.Conditions, types.DiskConditionTypeSchedulable)
+	if diskReadyCondition.Status != types.ConditionStatusTrue ||
+		diskSchedulableCondition.Status != types.ConditionStatusTrue {
+		return false, true
+	}
+	if !rcs.checkTagsAreFulfilled(diskSpec.Tags, v.Spec.DiskSelector) {
+		return false, true
+	}
+
+	return true, true
+}
+
+func (rcs *ReplicaScheduler) chooseReusableReplica(nodeInfo map[string]*longhorn.Node, rs map[string]*longhorn.Replica, nodeReusableReplicaMap map[string]*longhorn.Replica, volume *longhorn.Volume) (*longhorn.Replica, error) {
```

Comment:
Logic and flow here are similar to replica scheduling, except this is checking for an existing replica to see if it satisfied the current scheduling policy. We don't need to unify them now but it's better if we can unify the flow later so when we change the scheduling policy in the future, we won't need to modify two pieces of the code.